thirdeye-server/src/main/java/ai/startree/thirdeye/resources/CrudResource.java

```diff
@@ -206,11 +214,36 @@ public boolean hasAccess(final ThirdEyePrincipal principal, final AbstractDTO dt
 
   public void ensureHasAccess(final ThirdEyePrincipal principal, final AbstractDTO dto,
       final AccessType accessType) {
-    if (!hasAccess(principal, dto, accessType)) {
-      throw new ForbiddenException(Response.status(Status.FORBIDDEN).build());
+    ensureHasAccess(principal, ResourceIdentifier.fromDto(dto), accessType);
+  }
+
+  public void ensureHasAccess(final ThirdEyePrincipal principal, final ApiT api,
+      final AccessType accessType) {
+    ensureHasAccess(principal, ResourceIdentifier.fromDto(toDto(api)), accessType);
+  }
+
+  public void ensureHasAccess(final ThirdEyePrincipal principal, final ResourceIdentifier id,
+      final AccessType accessType) {
+    if (!accessControl.hasAccess(principal, id, accessType)) {
+      throw new ForbiddenException(Response.status(
+          Status.FORBIDDEN.getStatusCode(),
+          String.format("%s access denied for entity %s %s", accessType, id.entityType, id.name)
+      ).build());
     }
   }
 
+  public final void ensureHasAccessToRelatedResources(ThirdEyePrincipal principal, ApiT api) {
+    ensureHasAccessToRelatedResources(principal, toDto(api));
+  }
+
+  public final void ensureHasAccessToRelatedResources(ThirdEyePrincipal principal, DtoT dto) {
+    relatedDtos(dto).forEach(related -> ensureHasAccess(principal, related, AccessType.READ));
+  }
+
+  public List<ResourceIdentifier> relatedDtos(final DtoT dto) {
```

Comment:
nice! 

but this seems like an auth interface. need to see how to design this. The minus I see here is that Resource classes have to say how auth behaves in entities. don't think thats the best place for it.

Resource classes should ideally just collect data from services or push requests into classes which hold more of the business logic.